remote-activity: Add rows to general chart for remote realms.

laurynmm opened the pull request:
Updates the remote activity page for:
- Removes the "analytics users" column (commit 1)
- Adds a column for remote realms (ID, name, organization type) and a row is displayed in the chart for each remote realm on a remote server.
  - If the server doesn't have remote realms, then it still has a row displayed with any active plan, revenue and user information.
- Updates the remote activity key for these changes and to more clearly explain some of the existing chart data.

**Notes**:
- We were previously querying the database for each remote server to get the user counts. And adding the remote realms to the chart added two additional queries (one to get the remote realm object and then one to get the user counts)
  - Added functions to do a server and realm query for the audit logs to tally the user counts.
- We're still hitting the database for every current plan to get the revenue, so it would probably be good to see if we can get the latest license ledger with the plan query so that we can calculate a annual revenue estimate without so many queries.

@prakhar1144 and @alexmv - I'd appreciate your feedback on the database query updates in these commits. There might be some considerations around the audit logs that I haven't thought through. I'm a bit concerned where logs are bulk added with the same event time.

**Screenshots and screen captures:**

<details>
<summary>Updated remote activity view</summary>

![Screenshot from 2024-01-19 17-02-48](https://github.com/zulip/zulip/assets/63245456/ee3da08b-87b3-4960-9cb1-4e927022cf4e)

</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on analytics/views/remote_activity.py:
```diff
@@ -72,14 +72,20 @@ def get_remote_server_activity(request: HttpRequest) -> HttpResponse:
+    SERVER_HOSTNAME = 1
+    LAST_AUDIT_LOG_DATE = 4
+    MOBILE_USER_COUNT = 5
+    MOBILE_PUSH_COUNT = 6
```
This helped me with the different iterative changes on the for loops below, but there might be a better method for this. This also means that the column strings can be changed without breaking any logic below.

timabbott replied:
Yeah this seems better than what we had before!

laurynmm commented on corporate/lib/analytics.py:
```diff
@@ -80,9 +78,12 @@ def get_plan_data_by_remote_server() -> Dict[int, RemoteActivityPlanData]:  # no
-                current_status="Multiple plans",
-                current_plan_name="See support view",
+                current_status="ERROR: MULTIPLE PLANS",
+                current_plan_name=f"{current_plans}, {plan.name}",
```
This seemed better than the whole dataset not loading or something similar. Tried to make it something that would stand out.

laurynmm commented on analytics/tests/test_activity_views.py:
```diff
@@ -125,13 +161,31 @@ def test_activity(self, unused_mock: mock.Mock) -> None:
+            event_type=RemoteRealmAuditLog.USER_CREATED,
+            event_time=timezone_now() - timedelta(days=1),
+            extra_data=extra_data,
+        )
```
This just covers broad strokes, so it would be good to write a more complete/correct test data set up here.

timabbott commented on analytics/views/remote_activity.py:
```diff
@@ -78,28 +102,60 @@ def get_remote_server_activity(request: HttpRequest) -> HttpResponse:
+    plan_data_by_remote_realm = get_plan_data_by_remote_realm()
+
+    total_row = []
+    server_mobile_data_counted = []
```
This should be a set, right? Otherwise this algorithm is quadratic in the number of servers.

laurynmm replied:
Pushed update so this is a set and not a list. Also updated variable name to be clearer that it's remote servers.

timabbott commented on analytics/views/remote_activity.py:
```diff
@@ -78,28 +102,60 @@ def get_remote_server_activity(request: HttpRequest) -> HttpResponse:
+            plan_data = plan_data_by_remote_server.get(row[SERVER_ID])
+            user_counts = get_remote_server_guest_and_non_guest_count(row[SERVER_ID])
+        else:
+            server_remote_realms_data = plan_data_by_remote_realm.get(row[SERVER_ID])
```
Should this be called just `remote_realm_plan_data`? I'm not sure I understand the variable name.

laurynmm replied:
Renamed as suggested.

timabbott commented on corporate/lib/analytics.py:
```diff
@@ -147,3 +153,60 @@ def get_plan_data_by_remote_realm() -> Dict[int, Dict[int, RemoteActivityPlanDat
     return remote_server_plan_data_by_realm
+
+
+def get_realm_user_counts(
```
This should be `get_remote_realm_user_counts`, right?

laurynmm replied:
Renamed as suggested.

timabbott commented on corporate/lib/analytics.py:
```diff
@@ -147,3 +153,60 @@ def get_plan_data_by_remote_realm() -> Dict[int, Dict[int, RemoteActivityPlanDat
+        .select_related("server")
+    )
+
+    logs_per_server: Dict[int, List[RemoteRealmAuditLog]] = {}
```
You could use `defaultdict(list)` here and avoid the `if/else` for whether it's already empty; just writing `logs_per_server[log.server.id].append(log)`

laurynmm replied:
Thanks! I'd forgotten about `defaultdict`. Updated as such.

timabbott commented on corporate/lib/analytics.py:
```diff
@@ -147,3 +153,60 @@ def get_plan_data_by_remote_realm() -> Dict[int, Dict[int, RemoteActivityPlanDat
+        RemoteRealmAuditLog.objects.filter(
+            event_type__in=RemoteRealmAuditLog.SYNCED_BILLING_EVENTS,
+            event_time__lte=event_time,
+        )
```
The query itself looks correct.

We probably want to add a database index on `RemoteRealmAuditLog` to make these queries more efficient. This is probably easiest done testing in production.

That said, if we want to make this fast, probably the right path involves caching the data -- we store it in `LicenseLedger` for paid plans; I guess we could potentially maintain that data structure even for the base plans.

timabbott commented:
A few notes on the HTML layout itself:
* We don't need separate `ID` and `Realm ID` columns; realistically we may not need either? But it could be worth doing an "IDs" column with `server_id` or `server_id/realm_id` in it.
* Let's move "realm name" to the left, so it's next to the hostname.
* The "hostname" column should be `realm.host` for realm entries; that distinguishes them properly.
* We can rename the misleading columns like "Mobile pushes/Mobile users" to "Server mobile pushes" for now, if that lets us replace the documentation at the top.

laurynmm commented:
**Updates**:
- remote_server_mobile_data_counted is now a set, see https://github.com/zulip/zulip/pull/28634#discussion_r1459467393
- variable and function names clarified for these being remote server and remote realm data/actions.
- Used defaultdict(list) for remote server audit log data, see https://github.com/zulip/zulip/pull/28634#discussion_r1459487693
- Updated column names to indicate Server or Realm more clearly and deleted section in chart key about that

**TBD**:
- Moving the columns and popping data from the rows (for server and realm IDs) as suggesting in feedback above.

**Updated screenshot**

<details>
<summary>Remote activity view</summary>

![Screenshot from 2024-01-19 20-23-58](https://github.com/zulip/zulip/assets/63245456/743e4dbb-84c1-40ba-8dde-13551fea5929)

</details>

timabbott commented:
Sure, this looks like a great checkpoint, marked to merge once CI passes.
